copier/_main.py

```diff
@@ -1305,16 +1306,43 @@ def _apply_update(self) -> None:  # noqa: C901
             with local.cwd(subproject_top):
                 apply_cmd = git["apply", "--reject", "--exclude", self.answers_relpath]
                 ignored_files = git["status", "--ignored", "--porcelain"]()
-                # returns "!! file1\n !! file2\n"
-                # extra_exclude will contain: ["file1", file2"]
-                extra_exclude = [
-                    filename.split("!! ").pop()
-                    for filename in ignored_files.splitlines()
-                    if filename.startswith("!! ")
-                ]
-                for skip_pattern in chain(
-                    map(self._render_string, self.all_skip_if_exists), extra_exclude
-                ):
+                gitignore_path = Path(subproject_top, ".gitignore")
+                dir_patterns = set()
+                if gitignore_path.exists():
+                    with gitignore_path.open() as gitignore_file:
+                        for line in gitignore_file:
+                            line = line.strip()
+                            if not line or line.startswith("#"):
+                                continue
+                            # Matches foo/, foo/*, foo/**/, **/test/, **/test/*, etc.
+                            if (
+                                line.endswith("/")
+                                or line.endswith("/*")
+                                or line.endswith("/**/")
+                                or re.match(r".*\*\*/.*", line)
+                            ):
+                                cleaned = re.sub(r"(\/\*|\/\*\*\/|\/$)", "", line)
+                                dir_patterns.add(cleaned)
+                # Single loop: process ignored files and build exclude lists
+                ignored_dirs = set()
+                extra_exclude = []
+                for filename in ignored_files.splitlines():
+                    if filename.startswith("!! "):
+                        fname = filename.split("!! ").pop()
+                        matched_dir = False
+                        for p in dir_patterns:
+                            if fname == p or fname.startswith(p + "/"):
+                                if fname.endswith("/"):
+                                    ignored_dirs.add(fname.rstrip("/"))
+                                matched_dir = True
+                                break
+                        if not matched_dir:
+                            extra_exclude.append(fname)
+                for dir_pattern in ignored_dirs:
+                    apply_cmd = apply_cmd["--exclude", dir_pattern + "/*"]
+                for skip_pattern in map(self._render_string, self.all_skip_if_exists):
+                    apply_cmd = apply_cmd["--exclude", skip_pattern]
+                for skip_pattern in extra_exclude:
```

Comment:
TBH, this looks quite complex for a problem, I'm a bit worried that we're missing some edge cases.

I haven't looked into it in detail, but I think there are two ways to exclude files in the update:

1. Exclude files when applying the patch via `git apply --exclude`, which has the limitation you've encountered.
1. Omit the files to be excluded from the patch.

Option 2 sounds potentially simpler and more robust. WDYT?